Fix issue #52475: Make loop detector only consider reachable memory

brunocodutra opened the pull request:
As [suggested](https://github.com/rust-lang/rust/pull/51702#discussion_r197585664) by @oli-obk `alloc_id`s should be ignored by traversing all `Allocation`s in interpreter memory at a given moment in time, beginning by `ByRef` locals in the stack. 

- [x] Generalize the implementation of `Hash` for `EvalSnapshot` to traverse `Allocation`s 
- [x] Generalize the implementation of `PartialEq` for `EvalSnapshot` to traverse `Allocation`s 
- [x] Commit regression tests

Fixes #52626 
Fixes https://github.com/rust-lang/rust/issues/52849

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @petrochenkov (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

oli-obk commented:
I'll have a more detailed look later. Just wondering right now if we can reuse the StableHash stuff of incremental. Seems like we're trying to solve similar issues

brunocodutra commented:
> Just wondering right now if we can reuse the StableHash stuff of incremental.

I'm not sure I follow, but _reuse_ sounds good.

oli-obk commented:
Hmm alloc_id_recursion_tracker should be removed... noone uses it

So what I meant was reusing https://github.com/rust-lang/rust/blob/fa60b7212373a0984c014c2ada74a0538b5b324c/src/librustc/ich/impls_ty.rs#L392

petrochenkov commented:
r? @oli-obk

brunocodutra commented:
@ecstatic-morse that's a great point, in fact my idea is to crawl the transitive allocations in advance as part of constructing `EvalContextRef`, so that `Hash` and `Eq` are trivial.

brunocodutra commented:
My understanding of the internals of MIRI are still flake, so please bear with me if I'm talking nonsense, but after some testing, I'm growing skeptical that the exact example given in #52475 (which I reproduce below) can actually be detected, not because of `AllocId`s changing, but because the hidden local variables created by `&5` seem to imply `Frame::locals` is an ever growing list. In fact, if I ignore locals that are _not_`ByRef` in my implementation it does detect this case as a in infinite loop.

```
#![feature(const_fn, const_let)]

const fn churn_alloc_id() -> usize {
    let mut x: &i32 = &5;
    loop {
        x = &5;
    }
    0
}

fn main() {
    let _ = [(); churn_alloc_id()];
}
```

Does that make sense? 
If so, can anyone come up with an example where `AllocId`s keep changing, but the number of locals remain constant?

oli-obk commented:
> the hidden local variables created by &5 seem to imply Frame::locals is an ever growing list. 

Nope, rustc allocates a single variable and just reuses it.

if you look at the MIR of that function, you can see that it's even more extreme, because rustc generates a hidden static for the `&5`:

```rust
    let mut _0: usize;                   // return place
    scope 1 {
        let mut _1: &i32;                // "x" in scope 1 at src/main.rs:4:9: 4:14
    }
    scope 2 {
    }
    let mut _2: &i32;
    let mut _3: &i32;

    bb0: {                              
        StorageLive(_1);                 // bb0[0]: scope 0 at src/main.rs:4:9: 4:14
        StorageLive(_2);                 // bb0[1]: scope 0 at src/main.rs:4:23: 4:25
        _2 = promoted[1];                // bb0[2]: scope 0 at src/main.rs:4:23: 4:25
                                         // mir::Constant
                                         // + span: src/main.rs:4:23: 4:25
                                         // + ty: &i32
                                         // + literal: promoted[1]
        _1 = _2;                         // bb0[3]: scope 0 at src/main.rs:4:23: 4:25
        StorageDead(_2);                 // bb0[4]: scope 0 at src/main.rs:4:25: 4:26
        goto -> bb1;                     // bb0[5]: scope 1 at src/main.rs:5:5: 7:6
    }

    bb1: {                              
        StorageLive(_3);                 // bb1[0]: scope 1 at src/main.rs:6:13: 6:15
        _3 = promoted[0];                // bb1[1]: scope 1 at src/main.rs:6:13: 6:15
                                         // mir::Constant
                                         // + span: src/main.rs:6:13: 6:15
                                         // + ty: &i32
                                         // + literal: promoted[0]
        _1 = _3;                         // bb1[2]: scope 1 at src/main.rs:6:9: 6:15
        StorageDead(_3);                 // bb1[3]: scope 1 at src/main.rs:6:15: 6:16
        goto -> bb1;                     // bb1[4]: scope 1 at src/main.rs:5:5: 7:6
    }
```

if you change the code to

```rust
    let x = 5;
    let mut x: *const u32 = &x;
    loop {
        let y = 5;
        x = &y;
    }
    0
```

you get 

```
    let mut _0: usize;                   // return place
    scope 1 {
        let _1: u32;                     // "x" in scope 1 at src/main.rs:4:9: 4:10
        scope 3 {
            let mut _2: *const u32;      // "x" in scope 3 at src/main.rs:5:9: 5:14
            scope 5 {
                let _6: u32;             // "y" in scope 5 at src/main.rs:7:13: 7:14
            }
            scope 6 {
            }
        }
        scope 4 {
        }
    }
    scope 2 {
    }
    let mut _3: *const u32;
    let mut _4: &u32;
    let mut _5: &u32;
    let mut _7: *const u32;
    let mut _8: &u32;
    let mut _9: &u32;

    bb0: {                              
        StorageLive(_1);                 // bb0[0]: scope 0 at src/main.rs:4:9: 4:10
        _1 = const 5u32;                 // bb0[1]: scope 0 at src/main.rs:4:13: 4:14
                                         // ty::Const
                                         // + ty: u32
                                         // + val: Value(ByVal(Bytes(5)))
                                         // mir::Constant
                                         // + span: src/main.rs:4:13: 4:14
                                         // + ty: u32
                                         // + literal: const 5u32
        StorageLive(_2);                 // bb0[2]: scope 1 at src/main.rs:5:9: 5:14
        StorageLive(_3);                 // bb0[3]: scope 1 at src/main.rs:5:29: 5:31
        StorageLive(_4);                 // bb0[4]: scope 1 at src/main.rs:5:29: 5:31
        StorageLive(_5);                 // bb0[5]: scope 1 at src/main.rs:5:29: 5:31
        _5 = &_1;                        // bb0[6]: scope 1 at src/main.rs:5:29: 5:31
        _4 = _5;                         // bb0[7]: scope 1 at src/main.rs:5:29: 5:31
        _3 = move _4 as *const u32 (Misc); // bb0[8]: scope 1 at src/main.rs:5:29: 5:31
        _2 = move _3;                    // bb0[9]: scope 1 at src/main.rs:5:29: 5:31
        StorageDead(_3);                 // bb0[10]: scope 1 at src/main.rs:5:30: 5:31
        StorageDead(_4);                 // bb0[11]: scope 1 at src/main.rs:5:30: 5:31
        StorageDead(_5);                 // bb0[12]: scope 1 at src/main.rs:5:31: 5:32
        goto -> bb1;                     // bb0[13]: scope 3 at src/main.rs:6:5: 9:6
    }

    bb1: {                              
        StorageLive(_6);                 // bb1[0]: scope 3 at src/main.rs:7:13: 7:14
        _6 = const 5u32;                 // bb1[1]: scope 3 at src/main.rs:7:17: 7:18
                                         // ty::Const
                                         // + ty: u32
                                         // + val: Value(ByVal(Bytes(5)))
                                         // mir::Constant
                                         // + span: src/main.rs:7:17: 7:18
                                         // + ty: u32
                                         // + literal: const 5u32
        StorageLive(_7);                 // bb1[2]: scope 5 at src/main.rs:8:13: 8:15
        StorageLive(_8);                 // bb1[3]: scope 5 at src/main.rs:8:13: 8:15
        StorageLive(_9);                 // bb1[4]: scope 5 at src/main.rs:8:13: 8:15
        _9 = &_6;                        // bb1[5]: scope 5 at src/main.rs:8:13: 8:15
        _8 = _9;                         // bb1[6]: scope 5 at src/main.rs:8:13: 8:15
        _7 = move _8 as *const u32 (Misc); // bb1[7]: scope 5 at src/main.rs:8:13: 8:15
        StorageDead(_8);                 // bb1[8]: scope 5 at src/main.rs:8:14: 8:15
        _2 = move _7;                    // bb1[9]: scope 5 at src/main.rs:8:9: 8:15
        StorageDead(_7);                 // bb1[10]: scope 5 at src/main.rs:8:14: 8:15
        StorageDead(_9);                 // bb1[11]: scope 5 at src/main.rs:8:15: 8:16
        StorageDead(_6);                 // bb1[12]: scope 3 at src/main.rs:9:5: 9:6
        goto -> bb1;                     // bb1[13]: scope 3 at src/main.rs:6:5: 9:6
    }
```

Where you can see that it keeps taking the address of `_6` in `bb1`. There aren't infinite locals, just infinite allocations, because each `StorageDead(_6)` deletes the allocation, and each `StorageLive(_6)` creates a new one.

brunocodutra commented:
Awesome, thanks for the detailed explanation, I'll investigate later today
why then non`ByRef` locals seem to be preventing the detection, do you have
any clues?

oli-obk commented:
Most likely because `ByVal` can mean `Value::Scalar(Scalar::Ptr(ptr))` which in turn can again include an `AllocId`

oli-obk commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -121,31 +122,113 @@ impl<'mir, 'tcx: 'mir> PartialEq for Frame<'mir, 'tcx> {
+            for local in locals.iter() {
+                // Skip dead locals
+                if let Some(value) = local {
+                    match value {
```
I think you can just call `HashStable::hash_stable` on the value to obtain the same effect + simultaneously do the correct thing on `ByVal`. Have a look at the `HashStable` impl, it should look very much like your code.

brunocodutra replied:
So I tried reusing `HashStable` for value, however [rustc panics](https://github.com/rust-lang/rust/blob/master/src/librustc/ich/impls_ty.rs#L415):

```
thread 'main' panicked at 'no value for AllocId', libcore/option.rs:989:5
```

I guess that means dangling references aren't expected when computing the stable hash, but on the other hand I don't see why that would make the hash _not stable_. Indeed, hashing the potentially empty result directly does seem to fix the issue. 

Does that makes sense?

oli-obk replied:
Yea... That's probably wrong. We should not ICE on

```rust
#![feature(const_let)]

const FOO: *const u32 = {
    let x = 42;
    &x
};

fn main() {
    let z = FOO;
}
```

and similar code. (If we fixed the ICE on the above code, we'd just end up with the ICE you're seeing) So we can just treat dangling pointers as not existing by not hashing them.

brunocodutra replied:
Right, so if got what you're saying right, it's safe to fix the implementation of `HashStable` for `AllocId` such that it doesn't panic and then just reuse it to hash the `Frame::locals` here.

oli-obk replied:
yes

brunocodutra commented:
Oh I see now, `ByRef` vs `ByVal`, right, totally missed that.

brunocodutra commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -45,11 +45,8 @@ pub struct EvalContext<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
-    /// steps *until* the loop detector is enabled. When it is positive, it is
-    /// the number of steps after the detector has been enabled modulo the loop
-    /// detector period.
-    pub(crate) steps_since_detector_enabled: isize,
```
Right, this commit was part of an attempt to detect infinite recursion prior to reaching the stack size limit. The idea was to only snapshot the current frame, which would be much lighter weight, however I soon realized this would obviously lead to false positives and I abandoned the idea, but forgot to revert this commit before pushing this batch. 

In any case I have already reverted this on my local workspace.

oli-obk commented:
The implementation of HashStable uses the former, whereas the original infonite loop detection algorithm uses the latter, which one should I use?

Oh no. I forgot about this.

Most of the impls are on `StableHashingContext` instead of on a generic argument. I wonder how many of these impls actually need the context and could be generic instead.

brunocodutra commented:
> If we can make the impls generic, then we can just use `HashStable<Memory>`

I'm not sure this would be so straightforward, the allocation map seems to [come from a global variable](https://github.com/rust-lang/rust/blob/c7cba3d33f564be140275c4fb9e33c6dc2c97b21/src/librustc/ich/impls_ty.rs#L412) rather   than the hashing context, that is, `StableHashingContext` doesn't seem to require a way to resolve Allocations to be provided.

Based on that observation, my current implementation also resolves allocation using the global context, as opposed to `Memory::alloc_map`, but I wonder whether this is correct. What's the difference between this two mappings?

oli-obk commented:
The global context allocations are statics and constants, while the ones in the EvalContext are for computing the current static/constant.

cc @michaelwoerister do you think we can tune HashStable impls by making all the `HashStable<StableHashingContext>` impls be `HashStable<T>` with `T: HashingContext` which exposes a bunch of useful functions

```rust
trait HashingContext {
    fn with_allocation<T>(&self, f: impl FnOnce(&Allocation) -> T) -> T;
}
```

that allow reusing that code?

brunocodutra commented:
> The global context allocations are statics and constants, while the ones in the EvalContext are for computing the current static/constant.

I see, makes sense, it turns out we need to use both then.

This got me thinking, since the current infinite loop detection algorithm does not snapshot the static memory, I wonder whether it would produce false positives for loops that only change statics. I'll add a regression test for that too.

brunocodutra commented:
Also, according to my understanding, it is ok for us to reuse `hash_stable` as is for now, since the fact it ignores any non-static allocations only means that more hashing conflicts than necessary will occur, however, since the equality comparison of `EvalSnapdhot` will take both kinds of allocations into account, no false positives would be produced. This essentially means that generalizing `hash_stable` to also take non-static allocations into account would be an optimization that could be tracked on a follow up issue, do you think that makes sense @oli-obk?

brunocodutra commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -133,44 +133,43 @@ impl<'a, 'mir, 'tcx: 'mir> HashStable<StableHashingContext<'a>> for Frame<'mir,
 }
 
-impl<'a, 'b, 'mir, 'tcx, M> HashStable<StableHashingContext<'b>> for EvalSnapshot<'a, 'mir, 'tcx, M>
+impl<'a, 'mir, 'tcx, M> HashStable<StableHashingContext<'a>> for EvalSnapshot<'mir, 'tcx, M>
```
The macro seems to [hardcode the lifetime parameters](https://github.com/rust-lang/rust/blob/f498e4ec1b57c3245a2fb8e0d5d836ed56760d2d/src/librustc/macros.rs#L77), is there a straightforward way to extend it for multiple generic parameters?

michaelwoerister commented:
@oli-obk, can you give me some context on what you are trying to achieve?

brunocodutra commented:
@oli-obk do you think the implementation of `Eq` for `EvalSnapshot` I outlined above is correct?

From what you wrote it seems in addition to query `Memory` for `Allocation`s we might need to also fallback to the global context if no allocation is found?

oli-obk commented on src/librustc_mir/interpret/memory.rs:
```diff
@@ -74,11 +73,10 @@ impl<'a, 'mir, 'tcx, M> PartialEq for Memory<'a, 'mir, 'tcx, M>
 }
 
-impl<'a, 'mir, 'tcx, M> Hash for Memory<'a, 'mir, 'tcx, M>
+impl<'a, 'b, 'mir, 'tcx, M> HashStable<StableHashingContext<'b>> for Memory<'a, 'mir, 'tcx, M>
```
do we still need this impl now that we hash only the reachable memory?

brunocodutra replied:
Good question, I guess not, though we still need to hash `Memory::data` and `Memory::cur_frame`, however that can be done explicitly by the impl `HashStable` for `EvalSnapshot`, meaning probably can get rid of this impl.

brunocodutra commented:
@oli-obk I believe the fix is now complete, so I'm removing the [WIP] tag. I still need to create a PR for MIRI implementing `HashStable` for `Evaluator`, but it seems MIRI doesn't build against current rustc master.

As discussed, generalizing `StableHashingContext` as a trait that may be implemented for `Memory` was left out of scope and it's tracked by a separate issue #53084.

oli-obk commented on src/librustc_mir/interpret/snapshot.rs:
```diff
@@ -0,0 +1,304 @@
+
+/// The virtual machine state during const-evaluation at a given point in time.
+#[derive(Eq, PartialEq, Hash)]
+pub struct EvalSnapshot<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> {
```
I currently can't see what this scheme gains us over the previous one. It introduces a lot of complexity.

Doesn't your suggestion of using `StableHash` as a hash with loads of collisions and then using the equality operation to look at content instead of ids work?

brunocodutra replied:
That's still the idea exactly, the only thing I changed from the previous iteration was to pre-resolve `AllocId`s to something that can be compared equal by the time `EvalSnapshot` is constructed, so that that we can just derive `Eq` directly. The alternative is to unroll the equality comparison for `EvalSnapshot` manually, which turned out to be extremely complex as we saw a couple of iterations ago.

I had to create a new type for `AllocId` because I can't just reimplement `Eq`, since it has index semantics everywhere else.

oli-obk replied:
Hmm... If this level of complexity is necessary, I wonder if it's really worth doing this. We can never catch all cycles anyway.

I don't feel comfortable just adding more things to miri without a very clear benefit.

oli-obk replied:
> mapping the evaluation context to a data structure that is invariant to AllocId and may be trivially compared.

This definitely feels simpler, but it looks to me like in the end it also requires writing out all the crawling code.

brunocodutra replied:
Correct, that's what the trait `Snapshot` does, maybe you have an insight on a simpler way to implement it?

oli-obk replied:
I presume we'd need both

* the HashStableContext trait ( https://github.com/rust-lang/rust/issues/53084 )
* split out all the fields that should be part of the hash/eq into their own type and thus single field so we can derive everything on them

This way the only special code would be the one for the `HashStable` impl of `AllocId`

brunocodutra replied:
> split out all the fields that should be part of the hash/eq into their own type 

I'm not sure I understand your suggestion, could you elaborate?

To clarify my current approach, I created _snapshot_ versions of everything that can directly or indirectly contain an `AllocId` and the list is long: `Frame`, `Memory`, `LocalValue`, `Value`, `ScalarMaybeUndef`, `Scalar`, `Pointer`, `Allocation` and `Relocations`. Instances of this types form a [cyclic] graph that cannot easily be flattened into a simple structure, because many are enums whose variants contain other metadata besides the `AllocId`, such as _offsets_ and _alignments_, that must also be preserved and later compared for equality. The trait `Snapshot` that I introduced is very similar to the existing `HashStable` trait, except that instead of producing a hash, it clones the entire structure except for `AllocId`s, which are replaced by `AllocHashId`s.

oli-obk replied:
We would not end up comparing Memory. Instead we'd query Memory for its Allocation whenever we encounter an `AllocId`.

To support this we'd need a copy of `Memory`, so we can just clone the entire thing (let's ignore code that has memory leaks for now, as those would cause `Memory` to grow even if the amount of observable state does not grow)

Then the lhs of the comparison uses the current values and current memory object and the rhs of the comparison uses the old values and the clone of the old memory.

The EvalContext contains the Memory, but also a bunch of other things that we do not need (like the TyCtxt, the loop counters, ...) so everything but the irrelevant fields could be moved into a separate structure ("state" maybe?) and then we can clone, hash and compare these fields via their derived traits instead of writing everything by hand...

But maybe we do still need handwritten impls for equality, because comparisons have no state... so we'd end up with the Situation we have now... ugh..

Ok... so maybe your solution is the best one. I'll give it another review.

brunocodutra replied:
Right, the "state" you mention is `EvalSnapshot`, it needs to contain a _snapshot_ of `Machine`, the stack (`Vec<Frame>`) and `Memory`, where _snapshot_ here means a clone of the subset of fields that matter for the purpose of detecting repeated states.

Hashing this snapshot is the easy part, since the current implementation of `HashStable` sufficient (albeit not perfect). The problem is implementing `Eq`, because we can't just implement it for `AllocId` somehow, since we need `Memory` to be accessible by the time we compare them.

oli-obk replied:
> The problem is implementing Eq, because we can't just implement it for AllocId somehow, since we need Memory to be accessible by the time we compare them.

Yes, that's where I notices why your solution might be more elegant ;)

Sorry about the confused mess above

brunocodutra replied:
After thinking about the current approach versus the direct implementation of `PartialEq` for `EvalSnapshot` for a few days, I'm starting to think that despite the amount of *scary hand written code* required by the later, it's still better than the former:

1. The current approach has the serious issue that it still relies on hashing in order to implement `PartialEq`, which means it could potentially cause false positives and reject valid code. I've attempted several approaches to fix this over the past few days, but I wasn't able to come up with anything reasonable so far.
2. Since the equality comparison of `FrameSnapshot` and `MemorySnapshot` are independent of one another, we may spend precious CPU cycles comparing unreachable `Allocation`s.

Based on this I'm inclined to stash the current approach on a branch and give another shot at manually implementing `PartialEq` for `EvalSnapshot`. 

@oli-obk what do you think?

oli-obk replied:
Yes this seems sensible. Maybe you can build a helper macro similar to the ones for `HashStable` to reduce the amount of boilerplate?

RalfJung commented:
Actually, as far as hashing is concerned, could it make sense to just not hash memory? That will reduce a lot the hashing cost incurred by loop detection.

It also increases the likelihood of collisions, and certainly the main snapshot comparison cannot ignore memory -- but my guess would be that collisions are still unlikely, because the value of immediate locals (which will be many!) as well as the full stack (including return addresses) is still being hashed. So overall I would expect this to be better.

Either way, @brunocodutra could you check if your PR fixes the code in https://github.com/rust-lang/rust/issues/52849#issuecomment-410635683?

RalfJung commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -213,33 +185,32 @@ impl<'a, 'mir, 'tcx, M> Default for InfiniteLoopDetector<'a, 'mir, 'tcx, M>
             // No collision
             return Ok(())
         }
 
```
Could you add an `info!` logging here saying that we are taking a snapshot? Such a snapshot is anyway so expensive that the extra check whether we should log or not should not make a difference -- and this would help a lot when debugging mysterious CTFE slowdowns.

brunocodutra replied:
Sure.

brunocodutra commented:
@RalfJung regarding your suggestion to not hash `Memory` and rely on the stack to keep collisions low makes a lot of sense, in fact I had reached the same conclusion myself, but didn't get to push that change (I've been crazy busy in the past two weeks).

As for checking whether this fixes the perf regression, what's the correct way to benchmark this?

brunocodutra commented:
@oli-obk and @RalfJung  I rebased the changes onto the latest master, added a convenience macro to avoid code duplication while implementing the `Snapshot` trait and added a `info!` log when cloning the constant evaluation context. Could you guys review my changes and let me know if there's anything else I could do to improve it?

RalfJung commented on src/librustc_mir/interpret/eval_context.rs:
```diff
@@ -227,21 +201,23 @@ impl<'a, 'mir, 'tcx, M> InfiniteLoopDetector<'a, 'mir, 'tcx, M>
-        let mut fx = FxHasher::default();
-        snapshot.hash(&mut fx);
-        let hash = fx.finish();
+        let mut hcx = memory.tcx.get_stable_hashing_context();
```
Please use `self.tcx` unless there is a reason not to (and then document taht reason^^).

brunocodutra replied:
Alright.

RalfJung commented:
> what's the correct way to benchmark this?

This is a case of the compile time going from reasonable to non-terminating, so basically compile that code and if it takes <20sec, you are probably good (I try to make things take 10s on my machine but yours may be slower), if it takes >1min the regression is still present.

> I'm curious what you mean by not hashing Memory?

I had not seen that you have already done that. Yes I think this PR implements what I was asking for.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/420798947) ([raw log](https://api.travis-ci.org/v3/job/420798947/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:05:08] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:05:09] tidy error: /checkout/src/librustc_mir/interpret/place.rs:58: line longer than 100 chars
[00:05:09] tidy error: /checkout/src/librustc_mir/interpret/snapshot.rs:5: line longer than 100 chars
[00:05:09] tidy error: /checkout/src/librustc_mir/interpret/snapshot.rs:32: line longer than 100 chars
[00:05:09] tidy error: /checkout/src/librustc_mir/interpret/snapshot.rs:43: line longer than 100 chars
[00:05:09] tidy error: /checkout/src/librustc_mir/interpret/snapshot.rs:160: line longer than 100 chars
[00:05:09] tidy error: /checkout/src/librustc_mir/interpret/snapshot.rs:263: line longer than 100 chars
[00:05:09] tidy error: /checkout/src/librustc_mir/interpret/snapshot.rs:271: line longer than 100 chars
[00:05:09] tidy error: /checkout/src/librustc_mir/interpret/snapshot.rs:292: line longer than 100 chars
[00:05:09] tidy error: /checkout/src/librustc_mir/interpret/eval_context.rs:116: line longer than 100 chars
[00:05:10] some tidy checks failed
[00:05:10] 
[00:05:10] 
[00:05:10] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:05:10] 
[00:05:10] 
[00:05:10] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:05:10] Build completed unsuccessfully in 0:00:52
[00:05:10] Build completed unsuccessfully in 0:00:52
[00:05:10] make: *** [tidy] Error 1
[00:05:10] Makefile:79: recipe for target 'tidy' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:007cc307
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:106efc1c:start=1535302819249326528,finish=1535302819258508968,duration=9182440
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:03a32840
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:044d33cb
travis_time:start:044d33cb
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:039ad46c
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

brunocodutra commented:
@RalfJung I believe this PR does fix that perf regression. While I gave up on the latest nightly compiling that piece of code after a couple of minutes had elapsed, my custom build finishes in about 20s.

```
time rustc +stage1 ctfe.rs 
real    0m22.461s
user    0m21.120s
sys     0m1.301s
```

oli-obk commented:
This looks great to me now.

Can you address @RalfJung 's nit and add some module level documentation to `snapshot.rs` explaining the the process?

RalfJung commented on src/librustc/mir/interpret/mod.rs:
```diff
@@ -129,8 +129,8 @@ impl<T: layout::HasDataLayout> PointerArithmetic for T {}
-pub struct Pointer {
-    pub alloc_id: AllocId,
+pub struct Pointer<Id=AllocId> {
+    pub alloc_id: Id,
```
Maybe add a comment why this is generic, that's fairly non-obvious. :D

RalfJung commented:
All that snapshot-specific code in `eval_context.rs` -- the `Hash` and `PartialEq` instances -- could you move that to the new file as well?

Also, there will be some "fun" conflicts with https://github.com/rust-lang/rust/pull/53671 :(

bors commented:
:umbrella: The latest upstream changes (presumably #53671) made this pull request unmergeable. Please resolve the merge conflicts.

brunocodutra commented:
@oli-obk @RalfJung 

* Rebased on top of the latest master
* Moved `InfiniteLoopDetector` to `snapshot.rs`
* Added some basic documentation
* Fixed regression tests that had been "broken" by an assertion introduced in c141ccf158d8c660ef20a51104b701b4eb37822b

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/423456201) ([raw log](https://api.travis-ci.org/v3/job/423456201/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:05:04] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:05:04] tidy error: /checkout/src/librustc_mir/interpret/eval_context.rs:148: line longer than 100 chars
[00:05:05] some tidy checks failed
[00:05:05] 
[00:05:05] 
[00:05:05] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:05:05] 
[00:05:05] 
[00:05:05] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:05:05] Build completed unsuccessfully in 0:00:50
[00:05:05] Build completed unsuccessfully in 0:00:50
[00:05:05] make: *** [tidy] Error 1
[00:05:05] Makefile:79: recipe for target 'tidy' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:004852d2
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:12ec69a0:start=1535833906587594157,finish=1535833906595616449,duration=8022292
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:1e3c8a20
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:01c1f20e
travis_time:start:01c1f20e
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:1de643dc
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/423458837) ([raw log](https://api.travis-ci.org/v3/job/423458837/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:13:06]    Compiling rustc_metadata_utils v0.0.0 (file:///checkout/src/librustc_metadata_utils)
[00:13:06]    Compiling rustc_mir v0.0.0 (file:///checkout/src/librustc_mir)
[00:13:06]    Compiling rustc_typeck v0.0.0 (file:///checkout/src/librustc_typeck)
[00:13:07] error: expected outer doc comment
[00:13:07]   --> librustc_mir/interpret/snapshot.rs:22:1
[00:13:07]    |
[00:13:07] 22 | //! This module contains the machinery necessary to detect infinite loops
[00:13:07]    |
[00:13:07]    |
[00:13:07]    = note: inner doc comments like this (starting with `//!` or `/*!`) can only appear before items
[00:13:07] error: aborting due to previous error
[00:13:07] 
[00:13:07] error: Could not compile `rustc_mir`.
[00:13:07] 
[00:13:07] 
[00:13:07] Caused by:
[00:13:07]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name rustc_mir librustc_mir/lib.rs --color always --error-format json --crate-type dylib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -C metadata=7952c237ec0d4952 -C extra-filename=-7952c237ec0d4952 --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps --extern arena=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libarena-4ee92b74dcb65ca6.so --extern bitflags=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libbitflags-5a39798fe03e47f4.rlib --extern byteorder=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libbyteorder-8246be02936c9b1b.rlib --extern either=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libeither-0a515e87c8afea9e.rlib --extern graphviz=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libgraphviz-45ae4394366d07fd.so --extern log=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/liblog-c55d6c95192e4906.rlib --extern log_settings=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/liblog_settings-87ec950697a15ed0.rlib --extern polonius_engine=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libpolonius_engine-218f3033f29f5493.rlib --extern rustc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc-cfbc17aa3c766576.so --extern rustc_apfloat=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_apfloat-4a211d9e23f5aeb5.rlib --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-61336079186baa43.so --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-f882aab6100635ab.so --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-51fd1bd0441a9815.so --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-8d84add221c0f710.so --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-8d84add221c0f710.rlib --extern smallvec=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsmallvec-cb741677cd0e0351.rlib --extern syntax=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax-26b6009735d1b07c.so --extern syntax_pos=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax_pos-ef45b71e578357b1.so -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/backtrace-sys-ee16f6821aef40e9/out -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/miniz-sys-52504d5ed57fefc2/out` (exit code: 1)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

oli-obk commented on src/librustc_mir/interpret/snapshot.rs:
```diff
@@ -1,3 +1,17 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
```
copyright comments are not necessary anymore

brunocodutra replied:
Oh ok, I thought I had missed that the first time, I'll amend this evening.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 05cdf8dc3d490fbec2519de90b254f47ec7373e8 has been approved by `oli-obk`

<!-- @bors r=oli-obk 05cdf8dc3d490fbec2519de90b254f47ec7373e8 -->

bors commented:
:hourglass: Testing commit 05cdf8dc3d490fbec2519de90b254f47ec7373e8 with merge c3186914071b71af2c771ec0c0d044a1a8c215c2...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.8887), [status-travis](https://travis-ci.org/rust-lang/rust/builds/425375545?utm_source=github_status&utm_medium=notification)
Approved by: oli-obk
Pushing c3186914071b71af2c771ec0c0d044a1a8c215c2 to master...

brunocodutra commented:
Does stable hashing avoid cycles? I thought it did not and concluded one couldn't form cycles of allocations, at least not in const fn. Can you come up with a counter exemple that hangs the compiler?

RalfJung commented:
How would stable hashing avoid cycles...?

> Can you come up with a counter example that hangs the compiler?

I am sure this will be easy once we got a few more features stabilizied. Still trying to get one with the current restricted set.

---

Another thing: Why does `EvalSnapshot::new` copy *everything* with `clone`, and then later has a ridiculously expensive implementation of `PartialEq::eq` do compare? Wouldn't it be much better to do the work of `EvalSnapshot::snapshot` just once, in `EvalSnapshot::new`?

brunocodutra commented:
I refer to stable hashing because not only it's part of loop detection, but also pretty much how the snapshot itself is implemented.

As for the other question, I can't think of a good reason besides me being obtuse. I'll revisit this PR later this evening with a fresh mind and much better understanding of the language - this PR was the very first thing I ever wrote in Rust.

RalfJung commented:
> I refer to stable hashing because not only it's part of loop detection, but also pretty much how the snapshot itself is implemented.

Sure but it has nothing to do with detecting cycles.^^

And actually, it's not just cycles. Currently this duplicates allocations when they are referenced multiple times. So imagine
```
let x = 0;
let x = (&x, &x);
let x = (&x, &x);
let x = (&x, &x);
let x = (&x, &x);
let x = (&x, &x);
let x = (&x, &x);
let x = (&x, &x);
let x = (&x, &x);
```
This will duplicate the original assertion (just containing `0`) 2^n times. So, there definitely needs to be a deduplication. I will open an issue.

> As for the other question, I can't think of a good reason besides me being obtuse. I'll revisit this PR later this evening with a fresh mind and much better understanding of the language - this PR was the very first thing I ever wrote in Rust.

That's okay. :) I was just wondering if there is maybe something I missed. I am touching the snapshot stuff in https://github.com/rust-lang/rust/pull/54380, and I will see if I can change it to snapshot earlier and not copy all of memory. I won't do any deduplication though.

RalfJung commented:
Hm actually stable hashing does *something* with `tcx`. So maybe it has sufficient state to deduplicate something? I don't see where and how this interacts with the snapshotting TBH (at the very least, an extensive comment should be added).

RalfJung commented:
I give up. I tried making a snapshot early, but I cannot comprehend what this code does, and I really should get actual work done.^^

The problem I ran into is that the snapshot isn't actually a snapshot. It's not a deep copy, it's full of references to... somewhere. This definitely needs *lots* of documentation, right now, the code is incomprehensible and unmaintainable as far as I am concerned.

brunocodutra commented:
At a high level, what it does is to crawl up the allocation stack and
replace every weak reference (AllocId) by a strong shared reference, so
that a trivial equality comparison does the right thing, comparing the
actual allocations rather than their IDs.

I see now why we clone everything and use a reference at the time we
compare for equality: it's to avoid duplicating clones. Imagine the same
allocation is referred through its AllocId in several places. If we crawl
up the allocation structure and replace every AllocId by a clone of the
allocation, we will clone the same thing over and over again. Instead we
clone all allocations only once and then just map AllocIds to references by
the time we need to compare the state for equality. Does that make sense?
This is actually not super expensive, the equality comparison needs to
crawl up the structure either way, the only extra cost is querying Memory,
which is a tradr off to avoid duplicating clones.

Ralf Jung <notifications@github.com> schrieb am Do., 20. Sep. 2018, 11:43:

> I give up. I tried making a snapshot early, but I cannot comprehend what
> this code does, and I really should get actual work done.^^
>
> The problem I ran into is that the snapshot isn't actually a snapshot.
> It's not a deep copy, it's full of references to... somewhere. This
> definitely needs *lots* of documentation, right now, the code is
> incomprehensible and unmaintainable as far as I am concerned.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/52626#issuecomment-423115859>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AD2zj2IBljiJaKAKBwcV2VfjZHFGxGIPks5uc2NRgaJpZM4VaIrv>
> .
>

brunocodutra commented:
Don't worry :)

Ralf Jung <notifications@github.com> schrieb am Do., 20. Sep. 2018, 12:01:

> I opened #54384 <https://github.com/rust-lang/rust/issues/54384> to track
> the lack of documentation. It would be great if you could take care of that
> @brunoabinader <https://github.com/brunoabinader> while the memory of
> this PR is still fresh in your mind :)
>
> And sorry if I came across too complaining. I wanted to get some other
> work done and the loop detector made that so much more complicated, so I
> wanted to quickly fix it to be simpler and got frustrated when that didn't
> work and I couldn't figure out why. I shouldn't have loaded that
> frustration on you though. I am happy for your contributions!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/52626#issuecomment-423122139>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AD2zj_hKkOvtSAecSTxrfiVOl8Ip2iD9ks5uc2eBgaJpZM4VaIrv>
> .
>
